iot_config/src/admin_service.rs

```diff
@@ -0,0 +1,153 @@
+use crate::{
+    admin::{self, AuthCache, KeyType},
+    region_map::{self, RegionMap},
+    GrpcResult,
+};
+use anyhow::Result;
+use file_store::traits::MsgVerify;
+use futures::future::TryFutureExt;
+use helium_crypto::{Network, PublicKey};
+use helium_proto::{
+    services::iot_config::{
+        self, AdminAddKeyReqV1, AdminKeyResV1, AdminLoadRegionReqV1, AdminLoadRegionResV1,
+        AdminRemoveKeyReqV1,
+    },
+    Region,
+};
+use sqlx::{Pool, Postgres};
+use tonic::{Request, Response, Status};
+
+pub struct AdminService {
+    auth_cache: AuthCache,
+    pool: Pool<Postgres>,
+    region_map: RegionMap,
+    required_network: Network,
+}
+
+impl AdminService {
+    pub fn new(
+        auth_cache: AuthCache,
+        pool: Pool<Postgres>,
+        region_map: RegionMap,
+        required_network: Network,
+    ) -> Self {
+        Self {
+            auth_cache,
+            pool,
+            region_map,
+            required_network,
+        }
+    }
+
+    async fn verify_request_signature<R>(&self, request: &R) -> Result<(), Status>
+    where
+        R: MsgVerify,
+    {
+        self.auth_cache
+            .verify_signature(KeyType::Administrator, request)
+            .await
+            .map_err(|_| Status::permission_denied("invalid admin signature"))?;
+        Ok(())
+    }
+
+    fn verify_network(&self, public_key: PublicKey) -> Result<PublicKey, Status> {
+        if self.required_network == public_key.network {
+            Ok(public_key)
+        } else {
+            Err(Status::invalid_argument(format!(
+                "invalid network: {}",
+                public_key.network
+            )))
+        }
+    }
+
+    fn verify_public_key(&self, bytes: &[u8]) -> Result<PublicKey, Status> {
+        PublicKey::try_from(bytes)
+            .map_err(|_| Status::invalid_argument(format!("invalid public key: {bytes:?}")))
+    }
+}
+
+#[tonic::async_trait]
+impl iot_config::Admin for AdminService {
+    async fn add_key(&self, request: Request<AdminAddKeyReqV1>) -> GrpcResult<AdminKeyResV1> {
+        let request = request.into_inner();
+
+        self.verify_request_signature(&request).await?;
+
+        let key_type = KeyType::from_i32(request.key_type)
+            .map_err(|_| Status::invalid_argument("invalid key type supplied"))?;
+        let pubkey = self
```

Comment:
is this check not redundant, in that if the pub key is invalid wont it already be caught at the sig verification on line 75?

Reply:
Good question but these are checking different keys/things.
They function on line 75 is checking the request to add the new key was signed by an already-registered key (the default key in the db or any admin keys saved to the db) so it would fail if it attempted to validate the request with the key being submitted to the `add_key` method. 
this check validates the key you’re trying to register is a valid key worth saving and also converts the bytes to a PublicKey in order to do the follow up check that it’s a valid key for the network this config instance is servicing (mainnet vs devnet)